pkg/client/client.go

```diff
@@ -91,9 +93,22 @@ func (p ApplyClient) createOrUpdateObj(newResource runtime.Object, forceUpdate b
 
 	// retrieve the current 'resourceVersion' to set it in the resource passed to the `client.Update()`
 	// otherwise we would get an error with the following message:
-	// "nstemplatetiers.toolchain.dev.openshift.com \"basic\" is invalid: metadata.resourceVersion: Invalid value: 0x0: must be specified for an update"
+	// `nstemplatetiers.toolchain.dev.openshift.com "basic" is invalid: metadata.resourceVersion: Invalid value: 0x0: must be specified for an update`
 	originalGeneration := metaExisting.GetGeneration()
 	metaNew.SetResourceVersion(metaExisting.GetResourceVersion())
+
+	// also, if the resource to create is a Service and there's a previous version, we should retain its `spec.ClusterIP`, otherwise
+	// the update will fail with the following error:
+	// `Service "<name>" is invalid: spec.clusterIP: Invalid value: "": field is immutable`
+	switch newResource := newResource.(type) {
+	case *corev1.Service:
+		newResource.Spec.ClusterIP = existing.(*corev1.Service).Spec.ClusterIP
+	case *unstructured.Unstructured:
+		if clusterIP, found, err := unstructured.NestedString(existing.(*unstructured.Unstructured).Object, "spec", "clusterIP"); err == nil && found {
+			unstructured.SetNestedField(newResource.Object, clusterIP, "spec", "clusterIP")
```

Comment:
btw, this function returns an error:
```
Returns an error if value cannot be set because one of the nesting levels is not a map[string]interface{}.
```

Reply:
yes, but I checked the `err` and here we only have 1 level: `spec` which should not be anything else than `map[string]interface{}`. Or am I missing something?

Reply:
yeah, you are right - I didn't realize that. 
Anyway, wouldn't it be better to add either comment there explaining it (plus say linter to ignore it) or just handle it since it's very simple change?
```suggestion
			if err := unstructured.SetNestedField(newResource.Object, clusterIP, "spec", "clusterIP"); err != nil {
				return false, err
			}
```